Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,629 @@
+    another sampling method e.g. random without replacement, which, when calling `num_indices` samples is guaranteed to draw each index exactly once.  
+        """
+
+    def __init__(self, num_indices, function,  sampling_type=None, prob_weights=None):
+
+        self._type = sampling_type
+        self._num_indices = num_indices
+        self.function = function
+
+        if prob_weights is None:
+            prob_weights = [1/num_indices]*num_indices
+        else:
+            if abs(sum(prob_weights)-1) > 1e-6:
+                raise ValueError('The provided prob_weights must sum to one')
+
+            if any(np.array(prob_weights) < 0):
+                raise ValueError(
+                    'The provided prob_weights must be greater than or equal to zero')
+
+        self._prob_weights = prob_weights
+        self._iteration_number = 0
+
+    @property
+    def prob_weights(self):
+        return self._prob_weights
+
+    @property
+    def num_indices(self):
+        return self._num_indices
+
+    @property
+    def current_iter_number(self):
+        return self._iteration_number
+
+    def next(self):
+        """ 
+        Returns a sample from the list of indices `{0, 1, …, S-1}, where S is the number of indices and increments the sampler.
+        """
+
+        out = self.function(self._iteration_number)
+
+        self._iteration_number += 1
+        return out
+
+    def __next__(self):
+        return self.next()
+
+    def get_samples(self,  num_samples=20):
+        """
+        Returns the first `num_samples` produced by the sampler as a numpy array.
+
+        Parameters
+        ----------
+        num_samples: int, default = 20
+            The number of samples to return. 
+        """
+        save_last_index = self._iteration_number
+        self._iteration_number = 0
+
+        output = [self.next() for _ in range(num_samples)]
+
+        self._iteration_number = save_last_index
+
+        return np.array(output)
+
+    def __str__(self):
+        repres = "Sampler that selects from a list of indices {0, 1, …, S-1}, where S is the number of indices. \n"
+        repres += "Type : {} \n".format(self._type)
+        repres += "Current iteration number : {} \n".format(
+            self._iteration_number)
+        repres += "Number of indices : {} \n".format(self._num_indices)
+        repres += "Probability weights : {} \n".format(self._prob_weights)
+        return repres
+
+    @staticmethod
+    def sequential(num_indices):
+        """
+        Instantiates a sampler that outputs sequentially. 
+
+        Parameters
+        ----------
+        num_indices: int
+            One above the largest integer that could be drawn by the sampler. The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+        
+        Returns
+        -------
+        A Sampler  that can be called with Sampler.next()  or next(Sampler) and outputs sequentially
```

Comment:
Sampler
An instance of the Sampler class representing the desired configuration.
```suggestion
        Returns
        -------
        Sampler
            An instance of the Sampler class that will generate indices sequentially.
```